CFT Redis - initial import from internal

jasonmar opened the pull request:
This PR adds terraform modules used to Redis Cluster with Sentinel managed replication.
The default configuration deploys three Redis nodes, each with Sentinel service enabled.
Each node is deployed to its own Managed Instance Group.
A GCS bucket is used to store startup script and configuration files for each node.

morgante reviewed (changes requested):
We should not delete the CFT template.

Commit:
Initial commit

morgante commented on test/fixtures/redis_cluster/main.tf:
```diff
@@ -0,0 +1,61 @@
+  zone_b            = "us-east1-a"
+  zone_c            = "us-east1-a"
+  network_project   = var.project_id
+  network           = random_pet.main.id
```
This needs to be a reference to the google_compute_network resource above so it gets passed all the way through.

Ex. `google_compute_network.test.name`

Likewise for subnet.

jasonmar replied:
Thanks, passing in the name via this reference caused terraform to wait for the network to be ready.

naseemkullah commented:
Any chance of this getting merged sometime soon?

morgante commented on modules/admin/main.tf:
```diff
@@ -0,0 +1,103 @@
+}
+
+# Needed for CI to wait for Cloud Storage API to become ready
+resource "null_resource" "api" {
```
Can we put this in a fixture or even the Cloud Build config? Just somewhere besides the module itself, since it's only needed for CI.

jasonmar replied:
done

morgante commented on modules/admin/variables.tf:
```diff
@@ -14,10 +14,10 @@
-variable "bucket_name" {
-  description = "The name of the bucket to create"
-}
+variable "project_id" { default = "" }
```
We should never have empty/non-usable default values.

If the value is required (like `project_id` is), you should simply not provide a default at all.

naseemkullah replied:
Not applicable here per se, but an empty value as default for things like "network" or "subnetwork" default to the default network which is a pretty sane default.

morgante replied:
In those cases, we could use "default" as the default value.

jasonmar replied:
removed empty default values

morgante commented on modules/admin/variables.tf:
```diff
@@ -14,10 +14,10 @@
-  description = "The name of the bucket to create"
-}
+variable "project_id" { default = "" }
+variable "service_account" { default = "redis-server" }
```
Please add descriptions for all variables, as per go/terraform-standards.

jasonmar replied:
done

morgante commented on modules/redis/main.tf:
```diff
@@ -0,0 +1,118 @@
+  content = data.template_file.config.rendered
+}
+
+resource "google_compute_instance_template" "redis" {
```
Could we consider updating this to use https://github.com/terraform-google-modules/terraform-google-vm/tree/master/modules/instance_template and https://github.com/terraform-google-modules/terraform-google-vm/tree/master/modules/mig?

jasonmar replied:
I'll do this in a secondary PR

morgante replied:
SGTM

morgante commented on modules/redis/variables.tf:
```diff
@@ -0,0 +1,41 @@
+ * limitations under the License.
+ */
+
+variable "cluster_name" { type = string }
```
Please format these on separate lines. (and run `terraform fmt`)

jasonmar replied:
done

morgante commented on modules/redis/variables.tf:
```diff
@@ -0,0 +1,41 @@
+ */
+
+variable "cluster_name" { type = string }
+variable "project_id" { type = string }
```
Please add descriptions for all input variables.

jasonmar replied:
done

naseemkullah commented:
Does this PR only create Sentinel clusters? Does it also provide a way to create a standalone non-ha single redis instance?

morgante commented:
@jasonmar Do you think you'll have a chance to respond to the feedback, or should we assign to someone else?

naseemkullah commented:
Is there any other recommended approach to installing a redis VM in the interim?

Commit:
Addressed comments in PR #2
- move storage api enabled check to fixture
- add variable descriptions
- terraform fmt
- remove unused argument

jasonmar commented:
This template only creates Sentinel clusters.
